### PR TITLE
Crop a DensityMap using a particle set

### PR DESCRIPTION
**Pull request by saltzberg**
Allows cropping of an em:: DensityMap (such as a read-in .mrc file) using a
set of particles as the mask.

Voxels farther than the inputted distance threshold to all the particle
centers are set to a value of zero.

Setting "inverse" to false results in any voxels within the distance
threshold to any particles set to a value of zero.

### CODE REVIEW

**Comment by benmwebb** [comment]
@ichem001 can you review? I will also take a look.

File: modules/em/include/DensityMap.h [hunk 1]
```diff
@@ -373,6 +373,21 @@ class IMPEMEXPORT DensityMap : public IMP::Object {
       \return the new cropped map.
    */
   DensityMap *get_cropped(float threshold);
+  
+  //! Create and return a new cropped map based on particles
+  /** The map is cropped based on voxel proximity to a set of particles.
+      All voxel centers farther than [distance] away from any
+      particle center in [ps] will have their density value set to 0.0.
+      
+      setting inverse to true will set all voxel centeres within
```
Comment (benmwebb) [inline modules/em/include/DensityMap.h:382]: "centeres" is misspelled.

File: modules/em/include/DensityMap.h [hunk 2]
```diff
@@ -373,6 +373,21 @@ class IMPEMEXPORT DensityMap : public IMP::Object {
...
+      [distance] of a particle to 0.0.
+
+      \param[in] ps List of particles used to crop map
+      \param[in] distance Distance from particles at which to crop map
+      \param[in] inverse Set to true to crop the particle volume of the map
+      \return the new cropped map.
+   */
+  DensityMap *get_cropped(Particles ps, float distance, bool inverse = false);
```
Comment (benmwebb) [inline modules/em/include/DensityMap.h:390]: We generally (although not always) use `double` for floating-point arguments. There is little reason to use plain `float` on a modern machine.

File: modules/em/src/DensityMap.cpp [hunk 3]
```diff
@@ -1104,6 +1104,84 @@ DensityMap *DensityMap::get_cropped(const algebra::BoundingBox3D &bb) {
...
+DensityMap *DensityMap::get_cropped(Particles ps, float distance, bool inverse) {
+  
+  algebra::Vector3Ds coords;
+  algebra::Vector3D vox_cent;
+  float vox_dist; 
+  bool crop;
+  Pointer<DensityMap> cropped_map;
+
+  // Always good to have a list of coordinates
+  for (unsigned int i=0; i<ps.size(); i++){
```
Comment (benmwebb) [inline modules/em/src/DensityMap.cpp:1116]: Use `++i` rather than `i++`. In principle it is faster (although I suspect a modern compiler with optimize it out anyway) but either way, it indicates you're not using the original value of `i` for something.

File: modules/em/src/DensityMap.cpp [hunk 4]
```diff
@@ -1104,6 +1104,84 @@ DensityMap *DensityMap::get_cropped(const algebra::BoundingBox3D &bb) {
...
+    coords.push_back(core::XYZ(ps[i]).get_coordinates());
+  }
+  
+  if(inverse==true) {
+      // For an inverse map, we simply make a clone of the 
+      // current map, since the bounds will be the same
+      cropped_map = create_density_map(this);
+  } else {
+      // If we want a resultant map contining the particles
```
Comment (benmwebb) [inline modules/em/src/DensityMap.cpp:1125]: "containing" ?

File: modules/em/src/DensityMap.cpp [hunk 5]
```diff
@@ -1104,6 +1104,84 @@ DensityMap *DensityMap::get_cropped(const algebra::BoundingBox3D &bb) {
...
+        if(coords[i][0] < mincorner[0]) { mincorner[0]=coords[i][0];}
+        if(coords[i][0] > maxcorner[0]) { maxcorner[0]=coords[i][0];}
+        if(coords[i][1] < mincorner[1]) { mincorner[1]=coords[i][1];}
+        if(coords[i][1] > maxcorner[1]) { maxcorner[1]=coords[i][1];}
+        if(coords[i][2] < mincorner[2]) { mincorner[2]=coords[i][2];}
+        if(coords[i][2] > maxcorner[2]) { maxcorner[2]=coords[i][2];}
+      }
+
+      // build a bounding box and add the distance threshold
+      algebra::BoundingBox3D bb(mincorner, maxcorner);
```
Comment (benmwebb) [inline modules/em/src/DensityMap.cpp:1142]: There's already a constructor that takes a list of coordinates: https://integrativemodeling.org/2.14.0/doc/ref/classIMP_1_1algebra_1_1BoundingBoxD.html#a612cedf52eea6d53f4ce2d5e219541b7

If that doesn't work for you for some reason, you should add a comment explaining why here (or fix the constructor).

File: modules/em/src/DensityMap.cpp [hunk 6]
```diff
@@ -1104,6 +1104,84 @@ DensityMap *DensityMap::get_cropped(const algebra::BoundingBox3D &bb) {
...
+  }
+    
+  // Get number of voxels in the new map
+  long n_vox = cropped_map->get_number_of_voxels();
+  
+  // Loop over all voxels
+  for (long v=0; v<n_vox; v++){
+    vox_cent = cropped_map->get_location_by_voxel(v);    
+    
+    // crop is a bool to decide whther we zero this voxel
```
Comment (benmwebb) [inline modules/em/src/DensityMap.cpp:1156]: whether?

File: modules/em/src/DensityMap.cpp [hunk 7]
```diff
@@ -1104,6 +1104,84 @@ DensityMap *DensityMap::get_cropped(const algebra::BoundingBox3D &bb) {
...
+    // Default value of crop is opposite of inverse
+    // So inverse=true defaults to all voxels not cropped
+    // (crop=false)
+    crop=!inverse;
```
Comment (benmwebb) [inline modules/em/src/DensityMap.cpp:1160]: IMP coding style is to add whitespace around operators.

File: modules/em/src/DensityMap.cpp [hunk 8]
```diff
@@ -1104,6 +1104,84 @@ DensityMap *DensityMap::get_cropped(const algebra::BoundingBox3D &bb) {
...
+
+    // Check if any particle is within distance
+    for(unsigned int i=0; i<coords.size(); i++){
+      vox_dist = algebra::get_distance(vox_cent, coords[i]);
```
Comment (benmwebb) [inline modules/em/src/DensityMap.cpp:1164]: Use `get_squared_distance` instead and compare to the square of `distance`. Faster since you avoid a bunch of unnecessary square roots.

File: modules/em/src/DensityMap.cpp [hunk 9]
```diff
@@ -1104,6 +1104,84 @@ DensityMap *DensityMap::get_cropped(const algebra::BoundingBox3D &bb) {
...
+      
+      // If any particle is within distance, 
+      // we reverse our cropping decision and break
+      if(vox_dist < distance){ 
+	crop=inverse;
```
Comment (benmwebb) [inline modules/em/src/DensityMap.cpp:1169]: Indentation looks wrong here.

File: modules/em/test/test_crop.py [hunk 10]
```diff
@@ -35,7 +35,7 @@ def test_crop_map(self):
             self.assertAlmostEqual(IMP.algebra.get_distance(
                 scene_bb.get_corner(i),
                 cropped_bb.get_corner(i)), 0, 2)
-        cropped_scene.get_header().show()
+        #cropped_scene.get_header().show()
```
Comment (benmwebb) [inline modules/em/test/test_crop.py:38]: This seems unrelated to your new functionality, should be in a separate commit.

File: modules/em/test/test_crop.py [hunk 11]
```diff
@@ -80,6 +80,58 @@ def _test_crop_using_larger_extent(self):
...
+        self.assertAlmostEqual(IMP.algebra.get_distance(c_cent, ps_cent), 0.0, delta=1.0)
+
+        # The correlation between the ps and map should be better
+        cropped_cfs = 1-IMP.em.compute_fitting_score(ps, cropped_map)
+        full_cfs = 1-IMP.em.compute_fitting_score(ps, mrc)
+
+        self.assertGreater(cropped_cfs, full_cfs)
+
+        # The cropped map should have less density
+        self.assertLess(IMP.em.approximate_molecular_mass(cropped_map, 0.00), IMP.em.approximate_molecular_mass(mrc, 0.00))
```
Comment (benmwebb) [inline modules/em/test/test_crop.py:111]: Should wrap this line so it doesn't exceed 80 chars.

File: modules/em/test/test_crop.py [hunk 12]
```diff
@@ -80,6 +80,58 @@ def _test_crop_using_larger_extent(self):
...
+
+        # Inverse cropped map should have less density
+        self.assertLess(IMP.em.approximate_molecular_mass(inv_cropped_map, 0.00), IMP.em.approximate_molecular_mass(mrc, 0.00))
+
+        # Addition of the two cropped maps should be almost identical to the original
+        inv_cropped_map.add(cropped_map)
+
+        # This evaluates to greater than 1.0 for unknown reasons - bug in CCC?
+        ccc = IMP.em.CoarseCC.cross_correlation_coefficient(inv_cropped_map, mrc, 0)
+        self.assertGreater(ccc, 0.99)
```
Comment (benmwebb) [inline modules/em/test/test_crop.py:133]: You should open an issue if that's the case.

**Review by benmwebb — CHANGES_REQUESTED** [review]

**Comment by ichem001** [comment]
@saltzberg What is the difference between what you are trying to do and https://github.com/salilab/imp/blob/7ba7974c39808c2824eed0ac253eb1189f3a040c/modules/em/include/DensityMap.h#L579. 
You could create a density map based on a segment and mask everything else out - you will not run into issue with distance being radially symmetric for any given "r" - and will not change the box size which may cause issue if it changes.

**Comment by saltzberg** [comment]
> @saltzberg What is the difference between what you are trying to do and
> 
> https://github.com/salilab/imp/blob/7ba7974c39808c2824eed0ac253eb1189f3a040c/modules/em/include/DensityMap.h#L579
> 
> .
> You could create a density map based on a segment and mask everything else out - you will not run into issue with distance being radially symmetric for any given "r" - and will not change the box size which may cause issue if it changes.

I didn't notice that function, but I do think this is different.  My immediate use for the new function is to abstract the "tube" of volume that contains the backbone and sidechain density.  I'm not sure how one would manipulate creating the density to achieve that with the other function.  

Additionally, if you are trying to mask a density using high-resolution structure, I think it's more straightforward to use the atom positions rather than perform the additional step of creating a density from them.

I do agree that having a constant value for the distance threshold for all atoms is not ideal.  I could have it be a multiplier to the radius instead. That would additionally allow one to set a custom distance threshold for each atom (say for parts of the map that are more smeared than others).

File: modules/em/src/DensityMap.cpp (hunk 5 above)
Comment (saltzberg) [reply modules/em/src/DensityMap.cpp:1142]: Yes, since I found that it is easy to bad bounding box dimensions, I can just use that constructor.

**Comment by saltzberg** [comment]
> and will not change the box size which may cause issue if it changes.

What issues might you see will changing the box size?

I change the box size because I need to abstract very small pieces from a large map.  Cropping the box to only the portion abstracted also speeds up the algorithm by reducing the number of voxels that need to be assessed.

Would adding a flag to allow for maintaining the box size be useful? (e.g..: `keep_box_dimensions=true`)

**Comment by ichem001** [comment]
> I didn't notice that function, but I do think this is different. My immediate use for the new function is to abstract the "tube" of volume that contains the backbone and sidechain density. I'm not sure how one would manipulate creating the density to achieve that with the other function.

- Would take 3 lines of code in python to do it, corresponding to less than 3 for loops over the voxels:

1. sample density from particle at lower resolution (density extend further as resolution decrease) using the map you want to crop as input for the header (some version of this code should be very fast because transforming x, y, and z is independent for each of the three cooridnates). 
2. binarize the sampled density map using `IMP.em.binarize()`, where 1 where density exists, 0 therwise (one loop over voxels).
3. multiple the two maps (one loop over voxel).

 > Additionally, if you are trying to mask a density using high-resolution structure, I think it's more straightforward to use the atom positions rather than perform the additional step of creating a density from them.
> I do agree that having a constant value for the distance threshold for all atoms is not ideal. I could have it be a multiplier to the radius instead. That would additionally allow one to set a custom distance threshold for each atom (say for parts of the map that are more smeared than others).

- Using a sampled density map, the spread of the density is dependent on the radius of the particle and the resolution because of the forward model; that's the only reason why I suggested using the masking routine. Other than that, your function is a nice addition to the EM tool set.

File: modules/em/test/test_crop.py (hunk 12 above)
Comment (saltzberg) [reply modules/em/test/test_crop.py:134]: Here: https://github.com/salilab/imp/issues/1044

File: modules/em/src/DensityMap.cpp (hunk 7 above)
Comment (saltzberg) [reply modules/em/src/DensityMap.cpp]: e480611efd

**Comment by benmwebb** [comment]
@saltzberg if you think this is done, you need to say so... (you can use the "re-request review" button) then we can merge it.

File: modules/em/include/DensityMap.h [hunk 13]
```diff
@@ -373,6 +373,21 @@ class IMPEMEXPORT DensityMap : public IMP::Object {
...
+      
+      setting inverse to true will set all voxel centers within
+      [distance] of a particle to 0.0.
+
+      \param[in] ps List of particles used to crop map
+      \param[in] distance Distance from particles at which to crop map
+      \param[in] inverse Set to true to crop the particle volume of the map
+      \return the new cropped map.
+   */
+  DensityMap *get_cropped(Particles ps, double distance, bool inverse = false, bool keep_map_dimensions = false);
```
Comment (benmwebb) [inline modules/em/include/DensityMap.h:390]: You forgot to document the new `keep_map_dimensions` arg here.

File: modules/em/src/DensityMap.cpp [hunk 14]
```diff
@@ -1104,6 +1104,75 @@ DensityMap *DensityMap::get_cropped(const algebra::BoundingBox3D &bb) {
...
+  double dsq=distance*distance; // square of distance (for comparisons)
+  bool crop;
+  Pointer<DensityMap> cropped_map;
+
+  // Always good to have a list of coordinates
+  for (unsigned int i=0; i<ps.size(); ++i){
+    coords.push_back(core::XYZ(ps[i]).get_coordinates());
+  }
+  
+  if(inverse==true || keep_map_dimensions==true) {
```
Comment (benmwebb) [inline modules/em/src/DensityMap.cpp:1121]: `==true` is redundant for booleans. You can just say `if (inverse  || keep_map_dimensions) {`.

**Review by benmwebb — APPROVED** [review]
